In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os


In [2]:
# Create path variable
path = r'C:\Users\howel\OneDrive\Task 4\TASK 4\Task 4.4' 

# Import products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [3]:
# Create path variable
path = r'C:\Users\howel\OneDrive\Task 4\TASK 4\Task 4.4' 

# Import orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [4]:
# Create a dataframe
df_test = pd.DataFrame()

# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\howel\AppData\Local\Temp\ipykernel_8432\1293419150.py:9: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [5]:
df_test['mix'] = df_test['mix'].astype('str')

In [6]:
# Find missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [7]:
# Create a subset to view missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [8]:
# Create new dataframe without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

df_prods_clean.shape

(49677, 5)

In [9]:
# Look for full duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [10]:
# Create a new dataframe without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

df_prods_clean_no_dups.shape

(49672, 5)

In [11]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

I ran descriptive statistics on the df_ords dataframe. Looking at the minimum and maximum values for each column, everything appears to be consistent. For example, 'order_dow' (day of week) ranges from 0 to 6, and 'order_hour_of_day' ranges from 0 to 23. These are valid time ranges.

In [12]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

I ran the code to check for mixed-type columns. The code returned no column names, indicating that there are no mixed-type columns in the dataframe.

In [13]:
# Check for missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

The output shows that the column 'days_since_prior_order' has 206,209 missing values. This is likely because these missing values correspond to customers who are placing their very first order. If a user has an order number of 1, there cannot be a "prior order," so the days since the prior order would naturally be missing (NaN).

In [14]:
# Create a new dataframe to clean
df_ords_clean = df_ords

# Create a new column 'new_customer'
# If days_since_prior_order is NaN, it returns True (New Customer)
# If it is not NaN, it returns False (Returning Customer)
df_ords_clean['new_customer'] = df_ords['days_since_prior_order'].isnull() == True

# Check the dataframe to see the new column
df_ords_clean.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  new_customer  
0                  8                     NaN          True  
1                  7                    15.0         False  
2                 12                    21.0         False  
3                  7                    29.0         False  
4                 15                    28.0         False

I have chosen to create a new variable that acts as a flag based on the missing value. The missing values in 'days_since_prior_order' indicate that the customer is brand new (it is their first order). I created a new column called 'new_customer' which is set to True if the value is missing, and False otherwise. This allows us to easily identify new customers in our analysis.

In [15]:
df_ords_dups = df_ords[df_ords.duplicated()]
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, new_customer]
Index: []

I ran a duplicate check using the duplicated() function. No rows were returned, which indicates there are no full duplicates in the dataset.

Since no duplicates were found in the previous step, no action was required to address them.

In [16]:
# Export the clean products data
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index = False)

df_ords_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'), index = False)

In [17]:
# 1. Create a clean products dataframe by excluding missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

# 2. Create the final dataframe by dropping duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

df_prods_clean_no_dups.shape

(49672, 5)